platform/source/mbed_error.c

```diff
@@ -146,18 +149,34 @@ static mbed_error_status_t handle_error(mbed_error_status_t error_status, unsign
 
     //Clear the context capturing buffer
     memset(&current_error_ctx, 0, sizeof(mbed_error_ctx));
+
     //Capture error information
     current_error_ctx.error_status = error_status;
-    current_error_ctx.error_address = (uint32_t)caller;
     current_error_ctx.error_value = error_value;
+    if (error_status == MBED_ERROR_MEMMANAGE_EXCEPTION ||
+            error_status == MBED_ERROR_BUSFAULT_EXCEPTION ||
+            error_status == MBED_ERROR_USAGEFAULT_EXCEPTION ||
+            error_status == MBED_ERROR_HARDFAULT_EXCEPTION) {
+#if defined(__CORTEX_M)
+        mbed_fault_context_t *mfc = (mbed_fault_context_t *)error_value;
+        current_error_ctx.error_address = (uint32_t)mfc->PC_reg;
+        current_error_ctx.thread_current_sp = (uint32_t)mfc->SP_reg;
```

Comment:
"Thread current SP" would actually be `PSP`.

`SP_reg` will be either thread SP (PSP) if the exception happened in thread mode, or main SP (MSP) if the exception happened in handler mode. Second case is less likely but possible.

I think you want `PSP` specifically here, as you are going to show it as part of the "current thread" line, next to the pointer for the thread base, as if intending to compare them.

(Although it doesn't matter that much, as the fault handler should have printed all of SP, MSP and PSP anyway before entering us, right?)

Not sure the warning is the way to go - we have no portable non-M fault handler, so how could any using a Cortex-A "implement a non Cortex-M handler" in this file to silence it? I would just let it use the existing code, and skip the warning.

Reply:
@kjbracey-arm 

> I think you want `PSP` specifically here, 

Please note that there is an adjustment of PSP/MSP on [this assembly code](https://github.com/ARMmbed/mbed-os/blob/master/platform/source/TARGET_CORTEX_M/TOOLCHAIN_GCC/except.S#L157
). It looks like the assembly code is adjusting the PSP/MSP value offset caused by calling `Fault_Handler`. I experimented with gdb and made sure that `SP_reg` value is the exact same with `SP` register value just before the crash. If you observe [teraterm_20190826.1656.log](https://github.com/ARMmbed/mbed-os/files/3543865/teraterm_20190826.1656.log), you can see that `SP` and `PSP` are different.
```
[2019-08-26 16:57:01.354] SP   : 20002D30 <== correct value just before the crash
[2019-08-26 16:57:01.371] LR   : 000017DF
[2019-08-26 16:57:01.389] PC   : 000017C4
[2019-08-26 16:57:01.403] xPSR : 21000000
[2019-08-26 16:57:01.421] PSP  : 20002CC8 <== a bit tainted value due to Fault_Handler
[2019-08-26 16:57:01.438] MSP  : 2002FFC0
```

> (Although it doesn't matter that much, as the fault handler should have printed all of SP, MSP and PSP anyway before entering us, right?)

Yes. We print all the context as shown above. And we print this mode information as well.
```
[2019-08-26 16:57:01.571] Mode : Thread
[2019-08-26 16:57:01.586] Priv : Privileged
[2019-08-26 16:57:01.604] Stack: PSP
```

> I would just let it use the existing code, and skip the warning.

OK. I removed the warning in bdfef3f43fc8f102a285741e15975876d25ca318.

Thanks!

Reply:
Okay, that complicates a bit. So `SP` is a better value for "thread SP" than `PSP` if the exception was from thread mode. But if the exception was from handler mode, then `SP` is a better value for "handler SP" than `MSP`.

Then if you're looking for "thread stack pointer" specifically, then you'd want to take `SP` if from exception was from thread mode, or `PSP` if from handler mode. So you need to look at the `SPSEL` bit of the exception return value.

(I've been around all this stuff working on exception handling for pyOCD, which has to do the same sort of reconstruction logic, trying to cover all cases).

Reply:
@kjbracey-arm 

> Okay, that complicates a bit. So `SP` is a better value for "thread SP" than `PSP` if the exception was from thread mode. But if the exception was from handler mode, then `SP` is a better value for "handler SP" than `MSP`.

Yes. I believe `SP` is the better value in both cases. If that's what you meant.

In case of thread mode, `PSP` is provided as a source to `SP` via [except.S#L112](https://github.com/ARMmbed/mbed-os/blob/master/platform/source/TARGET_CORTEX_M/TOOLCHAIN_GCC/except.S#L112). In case of handler mode, `MSP` is provided as source to `SP` via [except.S#L107](https://github.com/ARMmbed/mbed-os/blob/master/platform/source/TARGET_CORTEX_M/TOOLCHAIN_GCC/except.S#L107). Our assembly already checks `EXC_RETURN` from [except.S#L110](https://github.com/ARMmbed/mbed-os/blob/master/platform/source/TARGET_CORTEX_M/TOOLCHAIN_GCC/except.S#L110).

For stack dump feature upcoming, I will check the mode and dump the right PSP/MSP. For this PR, I believe providing `SP` as it is a good choice.

Reply:
> In case of handler mode, `MSP` is provided as source to `SP` via [except.S#L107](https://github.com/ARMmbed/mbed-os/blob/master/platform/source/TARGET_CORTEX_M/TOOLCHAIN_GCC/except.S#L107).

So in that case `SP` is not a thread stack pointer at all. If you're not able to do adaptive behaviour here, then a fixed choice of `SP` is better to get precision for the common case of exception in thread mode, so it's the 1 to go for, even though it's worse for the rare case of exception in handler mode.

Reply:
Now I understand your point. You mean this printout
```
Current Thread: main  Id: 0x20000E50 Entry: 0x3D13 StackSize: 0x1000 StackMem: 0x20001E60 SP: 0x20002D30
```
should also be changed to correctly show the handler stack instead of the thread stack. I will test with handler mode crashes and see if I can improve today.

@0xc0170 , please hold off merging. 😊

Reply:
The current thread information is still useful in that case.

With the thread dumping info on (I'm not sure why it isn't by default), it would be the missing (and most important piece) of the full thread dump. The thread information extracted from the OS is only up-to-date for the non-active threads - we have to get the current thread's from the registers.

But if the crash is actually in handler mode, it could either be due to a random interrupt thing, totally unconnected to the current thread, or it could be due to something the current thread did (maybe enabling interrupts at a bad time, or some bug in a supervisor call to RTOS).

So it would make sense to show both current thread and handler mode info when crashing in handler mode.

Again - look at pyOCD, and this is basically how it displays info with its thread awareness.

It shows:

    * Current backtrace for current stack (by examining SP).
    * If in handler mode (main stack), also show thread and backtrace for process stack (by examining MSP).
    * If RTOS awareness, also show thread and backtrace for other non-current RTOS threads (by examining RTOS global state structures)

So the handler mode stack appears whenever stopped in handler mode, else you don't see it and only see threads.

This then ties in with https://github.com/mbedmicro/pyOCD/pull/430, which I really should get merged, which makes it show the aborting thread by default when stopping in the fault handler - at present, it shows you as having "SIGSEGV at fault handler", rather than the more logical "SIGSEGV at main thread location". The first is technically true at a hardware level, because it has entered the fault handler by the time the debugger gets there, but it's not the way that signals are supposed to be reported in GDB. (It's not as if your thing here is going to say hard fault at fault handler, is it?)

Reply:
@kjbracey-arm 
Please review the original description of this request.

> * Stack dump is enabled via `"platform.stack-dump-enabled": true` mbed_app.json configuration.
> * If stack dump is used in conjunction with `"platform.error-all-threads-info": true`, all the threads' stack will be dumped to provide holistic overview of all the threads.

Now with above two configs set to true, Mbed-OS will print the whole stack including the `handler mode`'s stack.

Also, note how `handler mode`'s printout is cleaned up like this.
```
Current Thread: <handler>  Id: 0x0 Entry: 0x0 StackSize: 0xE4 StackMem: 0x2002FF18 SP: 0x2002FF18 
```
There is no thread id in handler and there is no entry. Or if you know, please let me know.

Reply:
If you want a handler pseudo-thread (and that's a reasonable way to present it), it should be **in addition** to the current OS thread, rather than replacing it (like pyOCD). If I read the current code correctly, if I pass a bad pointer to `osMutexLock`, then we could enter the SVC handler, then hard fault inside it, and we wouldn't see the current thread that did that. We'd see "handler" and all other threads, and no backtrace showing the mutex lock at all.

(You should be able to test literally that scenario - do something like `osMutexLock((osMutexId_t) 0xDEADDEAD)`).

In the handler crash scenario, we don't know whether it's the handler's fault (eg bad interrupt handler) or the current thread's fault, like that mutex lock. So we should try to get both backtraces.